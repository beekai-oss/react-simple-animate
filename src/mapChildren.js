// @flow

export default function mapChildren(
  childrenStoreInState: Object | Array<any>,
  children: Object | Array<any>,
) {
  let previousIndex: number = 0;
  const result: Array<any> = [];
  const childrenShallowCopy = Array.isArray(children) ? children : [children];
  const childrenStoreInStateShallowCopy = Array.isArray(childrenStoreInState)
    ? childrenStoreInState
    : [childrenStoreInState];

  // any children got removed
  childrenStoreInStateShallowCopy.forEach((element: Object) => {
    if (!element.hasOwnProperty('key')) return;

    if (childrenShallowCopy.find(child => child.key === element.key)) {
      result.push(element);
    } else {
      result.push({ ...element, willUnmount: true });
    }
  });

  // append missing child
  childrenShallowCopy.forEach((child, index) => {
    if (child.hasOwnProperty('key') && !result.find(r => r.key === child.key)) {
      result.splice(index === 0 ? 0 : index + 1, 0, {
        ...child,
        willMount: true,
      });
    }

    previousIndex = index;
  });

  return result;
}
