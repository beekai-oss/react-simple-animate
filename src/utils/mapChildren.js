// @flow
import React from 'react';
import propsGenerator from './propsGenerator';
import type { Props, State } from '../animate';

const filterUnMountChildren = (children: Array<React$Element<any>>) =>
  children.filter((child: Object) => !child.willUnmount);

export default function mapChildren(props: Props, state: State) {
  const { childrenStoreInState, willLeave, willEnter } = state;

  if (!Array.isArray(childrenStoreInState) || !Array.isArray(props.children)) {
    return null;
  }

  let children = childrenStoreInState;

  if (willLeave) {
    children = filterUnMountChildren(childrenStoreInState);
  } else if (willEnter) {
    children = props.children;
  }

  return children.map((child: Object) => {
    if (!child) return null;

    let componentProps = {};
    let startAnimation = true;

    const { willMount = false, willUnmount = false } = child;
    const isAnimateComponent =
      child.type &&
      child.type.displayName &&
      child.type.displayName === 'ReactSimpleAnimate';

    if (isAnimateComponent) {
      if (willMount) {
        startAnimation = willEnter;
      } else if (willUnmount) {
        startAnimation = willLeave;
      }
    } else {
      componentProps = propsGenerator(
        {
          ...props,
        },
        state,
        {
          willUnmount,
          willMount,
        },
      );
    }

    return React.cloneElement(child, {
      ...{ ...child.props },
      ...(!isAnimateComponent && componentProps.style
        ? { style: componentProps.style }
        : { startAnimation }),
    });
  });
}
