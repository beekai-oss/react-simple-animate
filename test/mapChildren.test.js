import mapChildren from '../src/mapChildren';

describe('mapChildren', () => {
  it('should mark child is about to be unMount', () => {
    const elements = [
      {
        key: 1,
      },
      {
        key: 2,
      },
    ];
    const children = [
      {
        key: 1,
      },
    ];

    expect(mapChildren(elements, children)).toMatchSnapshot();
  });

  it('should append new children', () => {
    const elements = [
      {
        key: 1,
      },
      {
        key: 2,
      },
      {
        key: 3,
      },
    ];
    const children = [
      {
        key: 0,
      },
      {
        key: 1,
      },
      {
        key: 4,
      },
      {
        key: 6,
      },
      {
        key: 3,
      },
      {
        key: 5,
      },
    ];

    expect(mapChildren(elements, children)).toMatchSnapshot();
  });

  it('should work fine with object', () => {
    const elements = {
      key: 1,
    };
    const children = {
      key: 0,
    };
    expect(mapChildren(elements, children)).toMatchSnapshot();
  });

  it('should work correctly when noting changed', () => {
    const elements = {
      key: 1,
    };
    const children = {
      key: 1,
    };
    expect(mapChildren(elements, children)).toMatchSnapshot();
  });

  it('should work fine when all deleted', () => {
    const elements = {
      key: 1,
    };
    const children = {};
    expect(mapChildren(elements, children)).toMatchSnapshot();
  });

  it('should work fine by adding new child when children were empty', () => {
    const elements = {};
    const children = {
      key: 1,
    };
    expect(mapChildren(elements, children)).toMatchSnapshot();
  });

  it('should work fine with both empty', () => {
    const elements = {};
    const children = {};
    expect(mapChildren(elements, children)).toMatchSnapshot();
  });
});
