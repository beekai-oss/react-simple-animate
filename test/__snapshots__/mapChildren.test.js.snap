// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mapChildren should append new child 1`] = `
Array [
  Object {
    "key": 0,
    "willMount": true,
  },
  Object {
    "key": 1,
  },
  Object {
    "key": 2,
    "willUnmount": true,
  },
  Object {
    "key": 4,
    "willMount": true,
  },
  Object {
    "key": 6,
    "willMount": true,
  },
  Object {
    "key": 3,
  },
  Object {
    "key": 5,
    "willMount": true,
  },
]
`;

exports[`mapChildren should append new children 1`] = `
Array [
  Object {
    "key": 0,
    "willMount": true,
  },
  Object {
    "key": 1,
  },
  Object {
    "key": 2,
    "willUnmount": true,
  },
  Object {
    "key": 4,
    "willMount": true,
  },
  Object {
    "key": 6,
    "willMount": true,
  },
  Object {
    "key": 3,
  },
  Object {
    "key": 5,
    "willMount": true,
  },
]
`;

exports[`mapChildren should mark child is about to be unMount 1`] = `
Array [
  Object {
    "key": 1,
  },
  Object {
    "key": 2,
    "willUnmount": true,
  },
]
`;

exports[`mapChildren should work correctly when noting changed 1`] = `
Array [
  Object {
    "key": 1,
  },
]
`;

exports[`mapChildren should work fine by adding new child when children were empty 1`] = `
Array [
  Object {
    "key": 1,
    "willMount": true,
  },
]
`;

exports[`mapChildren should work fine when all deleted 1`] = `
Array [
  Object {
    "key": 1,
    "willUnmount": true,
  },
]
`;

exports[`mapChildren should work fine with both empty 1`] = `Array []`;

exports[`mapChildren should work fine with object 1`] = `
Array [
  Object {
    "key": 0,
    "willMount": true,
  },
  Object {
    "key": 1,
    "willUnmount": true,
  },
]
`;
